Fix spherical harm terminology swap

itellaetxe opened the pull request:
## Summary
For functions related to spherical harmonics, the terminology "degree" and "order" is often used backwards in docstrings and variable names. This PR ensures consistency when referring to "degree" and "order" in the code.
Closes Issue #2970 

## Main changes
- Standardized the symbols for the spherical harmonic concepts of order and degree with m_value and l_value, respectively. In case these are arrays, we use m_values and l_values to be more informative of the type of the variable.
- Changed sh_order to sh_degree_max as @ebrahimebrahim suggested in the issue.
- Updated the examples according to the new input argument names (sh_degree_max).

pep8speaks commented:
Hello @itellaetxe, Thank you for submitting the Pull Request !







* In the file [`dipy/direction/peaks.py`](https://github.com/dipy/dipy/blob/56651d6239d1bcc5313ea27227e4dd4e8790e876/dipy/direction/peaks.py):

> [Line 408:81](https://github.com/dipy/dipy/blob/56651d6239d1bcc5313ea27227e4dd4e8790e876/dipy/direction/peaks.py#L408): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 80 characters)

* In the file [`dipy/direction/tests/test_bootstrap_direction_getter.py`](https://github.com/dipy/dipy/blob/56651d6239d1bcc5313ea27227e4dd4e8790e876/dipy/direction/tests/test_bootstrap_direction_getter.py):

> [Line 90:81](https://github.com/dipy/dipy/blob/56651d6239d1bcc5313ea27227e4dd4e8790e876/dipy/direction/tests/test_bootstrap_direction_getter.py#L90): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (84 > 80 characters)
> [Line 165:81](https://github.com/dipy/dipy/blob/56651d6239d1bcc5313ea27227e4dd4e8790e876/dipy/direction/tests/test_bootstrap_direction_getter.py#L165): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (84 > 80 characters)
> [Line 197:81](https://github.com/dipy/dipy/blob/56651d6239d1bcc5313ea27227e4dd4e8790e876/dipy/direction/tests/test_bootstrap_direction_getter.py#L197): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (84 > 80 characters)





* In the file [`dipy/nn/histo_resdnn.py`](https://github.com/dipy/dipy/blob/56651d6239d1bcc5313ea27227e4dd4e8790e876/dipy/nn/histo_resdnn.py):

> [Line 55:81](https://github.com/dipy/dipy/blob/56651d6239d1bcc5313ea27227e4dd4e8790e876/dipy/nn/histo_resdnn.py#L55): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 80 characters)
> [Line 56:81](https://github.com/dipy/dipy/blob/56651d6239d1bcc5313ea27227e4dd4e8790e876/dipy/nn/histo_resdnn.py#L56): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 80 characters)
> [Line 182:81](https://github.com/dipy/dipy/blob/56651d6239d1bcc5313ea27227e4dd4e8790e876/dipy/nn/histo_resdnn.py#L182): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 80 characters)



* In the file [`dipy/reconst/csdeconv.py`](https://github.com/dipy/dipy/blob/56651d6239d1bcc5313ea27227e4dd4e8790e876/dipy/reconst/csdeconv.py):

> [Line 161:81](https://github.com/dipy/dipy/blob/56651d6239d1bcc5313ea27227e4dd4e8790e876/dipy/reconst/csdeconv.py#L161): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (84 > 80 characters)
> [Line 322:81](https://github.com/dipy/dipy/blob/56651d6239d1bcc5313ea27227e4dd4e8790e876/dipy/reconst/csdeconv.py#L322): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (93 > 80 characters)
> [Line 341:81](https://github.com/dipy/dipy/blob/56651d6239d1bcc5313ea27227e4dd4e8790e876/dipy/reconst/csdeconv.py#L341): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 80 characters)

* In the file [`dipy/reconst/forecast.py`](https://github.com/dipy/dipy/blob/56651d6239d1bcc5313ea27227e4dd4e8790e876/dipy/reconst/forecast.py):

> [Line 83:81](https://github.com/dipy/dipy/blob/56651d6239d1bcc5313ea27227e4dd4e8790e876/dipy/reconst/forecast.py#L83): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 80 characters)
> [Line 455:9](https://github.com/dipy/dipy/blob/56651d6239d1bcc5313ea27227e4dd4e8790e876/dipy/reconst/forecast.py#L455): [E741](https://duckduckgo.com/?q=pep8%20E741) ambiguous variable name 'l'
> [Line 473:9](https://github.com/dipy/dipy/blob/56651d6239d1bcc5313ea27227e4dd4e8790e876/dipy/reconst/forecast.py#L473): [E741](https://duckduckgo.com/?q=pep8%20E741) ambiguous variable name 'l'
> [Line 487:9](https://github.com/dipy/dipy/blob/56651d6239d1bcc5313ea27227e4dd4e8790e876/dipy/reconst/forecast.py#L487): [E741](https://duckduckgo.com/?q=pep8%20E741) ambiguous variable name 'l'

* In the file [`dipy/reconst/mapmri.py`](https://github.com/dipy/dipy/blob/56651d6239d1bcc5313ea27227e4dd4e8790e876/dipy/reconst/mapmri.py):

> [Line 1530:81](https://github.com/dipy/dipy/blob/56651d6239d1bcc5313ea27227e4dd4e8790e876/dipy/reconst/mapmri.py#L1530): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (91 > 80 characters)

* In the file [`dipy/reconst/mcsd.py`](https://github.com/dipy/dipy/blob/56651d6239d1bcc5313ea27227e4dd4e8790e876/dipy/reconst/mcsd.py):

> [Line 51:81](https://github.com/dipy/dipy/blob/56651d6239d1bcc5313ea27227e4dd4e8790e876/dipy/reconst/mcsd.py#L51): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (91 > 80 characters)
> [Line 228:81](https://github.com/dipy/dipy/blob/56651d6239d1bcc5313ea27227e4dd4e8790e876/dipy/reconst/mcsd.py#L228): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (88 > 80 characters)
> [Line 269:81](https://github.com/dipy/dipy/blob/56651d6239d1bcc5313ea27227e4dd4e8790e876/dipy/reconst/mcsd.py#L269): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (85 > 80 characters)
> [Line 491:81](https://github.com/dipy/dipy/blob/56651d6239d1bcc5313ea27227e4dd4e8790e876/dipy/reconst/mcsd.py#L491): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (91 > 80 characters)





* In the file [`dipy/reconst/tests/test_csdeconv.py`](https://github.com/dipy/dipy/blob/56651d6239d1bcc5313ea27227e4dd4e8790e876/dipy/reconst/tests/test_csdeconv.py):

> [Line 494:81](https://github.com/dipy/dipy/blob/56651d6239d1bcc5313ea27227e4dd4e8790e876/dipy/reconst/tests/test_csdeconv.py#L494): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (85 > 80 characters)
> [Line 688:81](https://github.com/dipy/dipy/blob/56651d6239d1bcc5313ea27227e4dd4e8790e876/dipy/reconst/tests/test_csdeconv.py#L688): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (85 > 80 characters)



* In the file [`dipy/reconst/tests/test_mcsd.py`](https://github.com/dipy/dipy/blob/56651d6239d1bcc5313ea27227e4dd4e8790e876/dipy/reconst/tests/test_mcsd.py):

> [Line 66:81](https://github.com/dipy/dipy/blob/56651d6239d1bcc5313ea27227e4dd4e8790e876/dipy/reconst/tests/test_mcsd.py#L66): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 80 characters)
> [Line 119:81](https://github.com/dipy/dipy/blob/56651d6239d1bcc5313ea27227e4dd4e8790e876/dipy/reconst/tests/test_mcsd.py#L119): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 80 characters)
> [Line 127:81](https://github.com/dipy/dipy/blob/56651d6239d1bcc5313ea27227e4dd4e8790e876/dipy/reconst/tests/test_mcsd.py#L127): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (84 > 80 characters)
> [Line 133:81](https://github.com/dipy/dipy/blob/56651d6239d1bcc5313ea27227e4dd4e8790e876/dipy/reconst/tests/test_mcsd.py#L133): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (85 > 80 characters)
> [Line 161:81](https://github.com/dipy/dipy/blob/56651d6239d1bcc5313ea27227e4dd4e8790e876/dipy/reconst/tests/test_mcsd.py#L161): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 80 characters)
> [Line 199:81](https://github.com/dipy/dipy/blob/56651d6239d1bcc5313ea27227e4dd4e8790e876/dipy/reconst/tests/test_mcsd.py#L199): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 80 characters)
> [Line 219:81](https://github.com/dipy/dipy/blob/56651d6239d1bcc5313ea27227e4dd4e8790e876/dipy/reconst/tests/test_mcsd.py#L219): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 80 characters)
> [Line 231:81](https://github.com/dipy/dipy/blob/56651d6239d1bcc5313ea27227e4dd4e8790e876/dipy/reconst/tests/test_mcsd.py#L231): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 80 characters)

* In the file [`dipy/reconst/tests/test_shm.py`](https://github.com/dipy/dipy/blob/56651d6239d1bcc5313ea27227e4dd4e8790e876/dipy/reconst/tests/test_shm.py):

> [Line 146:81](https://github.com/dipy/dipy/blob/56651d6239d1bcc5313ea27227e4dd4e8790e876/dipy/reconst/tests/test_shm.py#L146): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 80 characters)
> [Line 168:81](https://github.com/dipy/dipy/blob/56651d6239d1bcc5313ea27227e4dd4e8790e876/dipy/reconst/tests/test_shm.py#L168): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 80 characters)
> [Line 197:81](https://github.com/dipy/dipy/blob/56651d6239d1bcc5313ea27227e4dd4e8790e876/dipy/reconst/tests/test_shm.py#L197): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 80 characters)

* In the file [`dipy/tracking/tests/test_tracking.py`](https://github.com/dipy/dipy/blob/56651d6239d1bcc5313ea27227e4dd4e8790e876/dipy/tracking/tests/test_tracking.py):

> [Line 709:81](https://github.com/dipy/dipy/blob/56651d6239d1bcc5313ea27227e4dd4e8790e876/dipy/tracking/tests/test_tracking.py#L709): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (84 > 80 characters)











* In the file [`doc/examples/contextual_enhancement.py`](https://github.com/dipy/dipy/blob/56651d6239d1bcc5313ea27227e4dd4e8790e876/doc/examples/contextual_enhancement.py):

> [Line 197:81](https://github.com/dipy/dipy/blob/56651d6239d1bcc5313ea27227e4dd4e8790e876/doc/examples/contextual_enhancement.py#L197): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 80 characters)















* In the file [`doc/examples/reconst_mcsd.py`](https://github.com/dipy/dipy/blob/56651d6239d1bcc5313ea27227e4dd4e8790e876/doc/examples/reconst_mcsd.py):

> [Line 258:81](https://github.com/dipy/dipy/blob/56651d6239d1bcc5313ea27227e4dd4e8790e876/doc/examples/reconst_mcsd.py#L258): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 80 characters)

















Do see the [DIPY coding Style guideline](https://github.com/dipy/dipy/blob/master/doc/devel/coding_style_guideline.rst)

itellaetxe commented:
My linter was not showing me the E501 problem. Fixing it. Sorry

skoudoro commented:
Hi @itellaetxe,

We fixed the issue in master,  Can you rebase this branch or merge master to this branch?

Thanks

CHrlS98 commented:
Hi @skoudoro @itellaetxe @ebrahimebrahim,
The terminology swap was introduced in a previous PR of mine (#2206). We also briefly discussed the swap here https://github.com/dipy/dipy/pull/2206#pullrequestreview-484290448. However I now realize that what we called the "literature" at the time is actually limited to the diffusion MR literature and from a mathematician's/physicist's point-of-view it might be wrong. Still, both bases we support in DIPY (`descoteaux07` and `tournier07`) use the term _order_ to describe $\ell$ in the cited papers.

If we do proceed with this change we might want to add a note somewhere in the doc clearly explaining the issue. It looks like the dMRI community have settled on the terminology **order** and all the papers I came across using CSD reconstruction report the maximum SH **order** when they talk about $\ell_{max}$ (although it is true that the maximum SH order and degrees will be equal). 

This is very confusing and unclear to me why there is a swap between the dMRI literature and wikipedia. I'll also tag @karanphil here who might have something to add.

karanphil commented:
> Hi @skoudoro @itellaetxe @ebrahimebrahim, The terminology swap was introduced in a previous PR of mine (#2206). We also briefly discussed the swap here https://github.com/dipy/dipy/pull/2206#pullrequestreview-484290448. However I now realize that what we called the "literature" at the time is actually limited to the diffusion MR literature and from a mathematician's/physicist's point-of-view it might be wrong. Still, both bases we support in DIPY (`descoteaux07` and `tournier07`) use the term _order_ to describe ℓ in the cited papers.
> 
> If we do proceed with this change we might want to add a note somewhere in the doc clearly explaining the issue. It looks like the dMRI community have settled on the terminology **order** and all the papers I came across using CSD reconstruction report the maximum SH **order** when they talk about ℓmax (although it is true that the maximum SH order and degrees will be equal).
> 
> This is very confusing and unclear to me why there is a swap between the dMRI literature and wikipedia. I'll also tag @karanphil here who might have something to add.

It is indeed confusing. I remember being confused at first, coming from a physics background. However, when you read the literature on SH bases and CSD in dMRI (like tournier04, tournier07, descoteaux07 or descoteaux09), they all refer to `l` (or `n` in the case of Tournier/MRtrix) as the order of the SH function.

While Wikipedia follows the inverse terminology, other references such as Brilliant (https://brilliant.org/wiki/spherical-harmonics/) follow the current way of Dipy and the dMRI literature.

I think we need to be very careful if changing this.

skoudoro commented:
Thank you for your feedback @CHrlS98. Indeed, Looking back https://github.com/dipy/dipy/pull/2206 reminds me that I have already this question as a comment.

@Garyfallidis and @arokem, Can you also give your opinion here? it is important. thank you! 

maybe @RafaelNH could provide also a feedback

ebrahimebrahim commented:
> It looks like the dMRI community have settled on the terminology **order** and all the papers I came across using CSD reconstruction report the maximum SH **order** when they talk about $\ell_\text{max}$ (although it is true that the maximum SH order and degrees will be equal).

Hmm this is more confusing that I thought when I opened #2970

Sorry if I opened a can of worms that should have remained closed :grimacing: 

(In mrtrix "order" is used for $\ell$ but then it's not "degree" but "_phase_" that is used for $m$. So while using "order" for $\ell$ could match some dmri community conventions, I am questioning whether _both_ $\ell$=order _and_ $m$=degree is ever used together. IDK anymore...)

itellaetxe commented:
> We fixed the issue in master, Can you rebase this branch or merge master to this branch?

Hi @skoudoro, due to the ongoing discussion, I think I will leave it like this until we arrive to a conclusion. I will adapt the code to the final decision.

arokem commented:
Hi! I think that we should use the terminology that is used in the dMRI papers. I would suggest putting in a Note that explains that this is a choice that we made, so readers of wikipedia and scipy should be careful when relying on these as resources. Does that makes sense?

ebrahimebrahim commented:
Do dMRI papers swap degree and order or do they just use order for $\ell$ and then something different (like phase) for m?

CHrlS98 commented:
In Maxime Descoteaux (@mdesco) PhD thesis, there is a chapter on spherical harmonics (section 5.2) where it is said that $\ell$ is the order and $m$, the degree (https://theses.hal.science/tel-00457458):
![image](https://github.com/dipy/dipy/assets/41654474/a4a34314-bfff-4f61-b573-60d4938178d6)
I am however having a hard time finding a peer-reviewed journal article referring to $m$ as the degree.

In Tournier *et al*, 2004 (https://doi.org/10.1016/j.neuroimage.2004.07.037), they refer to $\ell$ as the order, but $m$ is indeed called a phase factor:
> Each spherical harmonic can be denoted using two numbers: its harmonic order n (n ≥ 0) and phase factor m (−n ≤ m ≤ n).

Actually, even in the DIPY paper from 2014 (https://doi.org/10.3389/fninf.2014.00008), $m$ is called the phase:
> The SH of order l and phase m, Yml(θ, ϕ), arises from the angular solution to Laplace's equation in spherical coordinates and they form an orthonormal basis for complex functions defined on the unit sphere. 

So, while there appears to be a consensus on $\ell$, I'm not really sure what we should call $m$ anymore...

arokem commented:
@desco also used the term "phase factor" in [his seminal 2007 paper]:(https://onlinelibrary.wiley.com/doi/pdf/10.1002/mrm.21277)

<img width="494" alt="Screenshot 2024-03-04 at 8 30 25 PM" src="https://github.com/dipy/dipy/assets/118582/dab95a79-940a-462b-ab71-5b90462effb4">

I think part of the confusion stems from the fact that the words "order" and "degree"  are used interchangeably to refer to the highest exponent in a polynomial function (see e.g., [here](https://math.stackexchange.com/questions/3323294/degree-and-order-of-a-polynomial)). So, a nomenclature that uses "degree" and "phase factor" is potentially much less confusing. In fact, one could argue that if you ever use "order" that should refer to the variable referred to as $\ell$ in this writing of the equation. 

In other words, potentially with a lot of explanation, I might suggest to go with `order` and `phase_factor` (I think that just `phase` may be confusing, because it's an integer and phases are usually thought of as continuous things.

skoudoro commented:
> In other words, potentially with a lot of explanation, I might suggest to go with order and phase_factor (I think that just phase may be confusing, because it's an integer and phases are usually thought of as continuous things.

+1 with this conclusion. Thanks all for the debate / clarification, I think we have a consensus.

@itellaetxe, can you go ahead with  `order` (for $\ell$) and `phase_factor / phase` for $m$  in the code. Concerning the documentation (rst file), you can refer to this PR for the discussion and add some additional note.

I plan to release DIPY tomorrow or in 2 days so it will be great if you could update this PR on ASAP. Thank you in advance!

ps: it is ok if you can not make it.

skoudoro commented:
PS: do not forget to fix the conflict and pep8 issues above. Thanks again !

itellaetxe commented:
Thanks to everyone that took part in the discussion. Shortly:

- I followed the indications given by @arokem and @skoudoro, regarding the new namings.
- `sh_degree_max` is now `sh_order_max`.
- Included the mentioned note in the theory rst file explaining our convention choice.
- Fixed a couple of bad namings I overlooked the first time I opened the PR.
- Passed tests locally.
- Rebased my branch onto master.


I just saw that I made a typo with `order` in a couple of locations. Fixing and pushing now.

skoudoro commented on doc/theory/sh_basis.rst:
```diff
@@ -161,6 +160,15 @@ are also taken into account when reconstructing a spherical function. These
 full bases can successfully reconstruct asymmetric signals as well as
 symmetric signals.
 
+NOTE: 
+The definition of spherical harmonics that DIPY utilizes does not match the one
+in Wikipedia and scipy. Instead, DIPY follows the dMRI literature conventions,
+like in ``descoteaux07`` and ``tournier07``.
+The code in DIPY also follows the following convention:
+Let the SH be noted as $Y_{l}^m$. Then, $l$ is referred to as either order or
+l_value(s), and $m$ is referred to as either phase factor or m_value(s).
+This decisions were made as a result of the PR in https://github.com/dipy/dipy/pull/3086
+
```
Perfect !

karanphil commented:
I am questioning the change from `sh_order` to `sh_order_max`. This would require at least deprecated versions for all functions with `sh_order=` as optional argument. For instance, `reconst.shm.sf_to_sh` will break now for any users using `sh_order=`.

I understand `sh_order_max` is more accurate, but is it really a necessary change?

skoudoro commented:
> I understand sh_order_max is more accurate, but is it really a necessary change?

Let's try to be the most accurate as possible but you got a very good point @karanphil 

@itellaetxe, everytime you change from `sh_order` to `sh_order_max` in a function argument, you need to add the decorator below:

```python
@deprecated_params('sh_order', 'sh_order_max', since='1.9', until='2.0')
```

see an example: https://github.com/dipy/dipy/blob/379062e3704088bcc998e3736a52212ca41627f5/dipy/align/imaffine.py#L249

karanphil commented on doc/theory/sh_basis.rst:
```diff
@@ -161,6 +160,15 @@ are also taken into account when reconstructing a spherical function. These
+The code in DIPY also follows the following convention:
+Let the SH be noted as $Y_{l}^m$. Then, $l$ is referred to as either order or
+l_value(s), and $m$ is referred to as either phase factor or m_value(s).
+This decisions were made as a result of the PR in https://github.com/dipy/dipy/pull/3086
```
Since this PR is cited there, I would also update the section "Main changes" (at the beginning of the PR) to represent the actual final changes. It would be less confusing if someone looks at it in the future.

itellaetxe replied:
Thank you both for the suggestions :)

- @karanphil I just updated the beginning of the PR. It should reflect the current changesnow.
- @CHrlS98 thanks for the grammar check. I will leave it as _"These decisions..."_. Pushing it in the next commit.

itellaetxe commented:
> @itellaetxe, everytime you change from `sh_order` to `sh_order_max` in a function argument, you need to add the decorator below:

It should be good to go now!

karanphil commented on dipy/data/__init__.py:
```diff
@@ -261,7 +261,7 @@ def mrtrix_spherical_functions():
     func_coef : array (2, 3, 4, 45)
         Functions represented by the coefficients associated with the
-        mrtrix spherical harmonic basis of order 8.
+        mrtrix spherical harmonic basis of order (l) 8.
```
Should it be "basis of maximal order"? Following the change from `sh_order` to `sh_order_max`.

karanphil commented on dipy/reconst/csdeconv.py:
```diff
@@ -150,15 +150,15 @@ def __init__(self, S0, dwi_response, bvalue=None):
-        self.n = np.arange(0, self.sh_order + 1, 2)
+        self.m_value = np.zeros(len(dwi_response))
+        self.sh_order_max = 2 * (len(dwi_response) - 1)
+        self.l_value = np.arange(0, self.sh_order_max + 1, 2)
```
`self.l_value` and `self.m_value` are arrays here, shouldn't they have `s` at the end?

karanphil commented on dipy/reconst/forecast.py:
```diff
@@ -79,8 +80,8 @@ def __init__(self,
-        sh_order : unsigned int,
-            an even integer that represent the SH order of the basis (max 12)
+        sh_order_max : unsigned int,
+            an even integer that represent the SH order (l) of the basis (max 12)
```
maximal SH order

karanphil commented on dipy/reconst/mcsd.py:
```diff
@@ -83,69 +86,70 @@ def __init__(self, response, sh_order, shells, S0=None):
-    n : int ``>= 0``
-        The degree of the harmonic.
+    sh_order_max : int ``>= 0``
+        The order (l) of the harmonic.
```
maximal order

karanphil commented on dipy/reconst/shm.py:
```diff
@@ -162,21 +163,22 @@ def gen_dirac(m, n, theta, phi, legacy=True):
-        The degree of the harmonic.
-    n : int ``>= 0``
-        The order of the harmonic.
+    m_value : int ``|m| <= l``
```
I believe that most of the `m_value` and `l_value` in `reconst.shm` are arrays. The old docstring might be erroneous. It might be "array of int" instead of "int".

karanphil commented on dipy/reconst/shm.py:
```diff
@@ -409,33 +411,33 @@ def real_sh_tournier(sh_order, theta, phi,
            NeuroImage. 2019 Nov 15;202:116-137.
     """
-    m, n = sph_harm_ind_list(sh_order, full_basis)
+    m_value, l_value = sph_harm_ind_list(sh_order_max, full_basis)
```
These are arrays I believe.

karanphil commented on dipy/reconst/shm.py:
```diff
@@ -747,26 +755,28 @@ def sampling_matrix(self, sphere):
-        sh_order : even int >= 0
-            the spherical harmonic order of the model
+        sh_order_max : even int >= 0
+            the spherical harmonic order (l) of the model
```
maximal

karanphil reviewed (commented):
I noted a couple of places where `l/m_value` should be `l/m_values`, and where the docstring is incomplete. Other than that, it looks good to me, great job!

skoudoro commented on dipy/reconst/csdeconv.py:
```diff
@@ -463,17 +466,17 @@ def estimate_response(gtab, evals, S0):
 
 
-def forward_sdt_deconv_mat(ratio, n, r2_term=False):
+def forward_sdt_deconv_mat(ratio, l_values, r2_term=False):
```
here, it is missing `@deprecated_params('n', 'l_values', since='1.9', until='2.0')

skoudoro commented on dipy/reconst/shm.py:
```diff
@@ -56,18 +57,18 @@ def bandit(f):
 
 
-def forward_sdeconv_mat(r_rh, n):
+def forward_sdeconv_mat(r_rh, l_values):
```
it is missing `@deprecated_params('n', 'l_values', since='1.9', until='2.0')`

skoudoro commented on dipy/reconst/shm.py:
```diff
@@ -201,42 +203,42 @@ def spherical_harmonics(m, n, theta, phi, use_scipy=True):
                         'dipy.reconst.shm.real_sh_descoteaux_from_index '
                         'instead', since='1.3', until='2.0')
-def real_sph_harm(m, n, theta, phi):
+def real_sph_harm(m_values, l_values, theta, phi):
```
same here

skoudoro commented on dipy/reconst/shm.py:
```diff
@@ -245,33 +247,33 @@ def real_sph_harm(m, n, theta, phi):
 
 
-def real_sh_tournier_from_index(m, n, theta, phi, legacy=True):
+def real_sh_tournier_from_index(m_values, l_values, theta, phi, legacy=True):
```
same here, see above

itellaetxe commented:
@skoudoro, I completely overlooked the deprecator decorator in the `m` and `n` cases. I will fix it. Sorry!

skoudoro commented on dipy/reconst/shm.py:
```diff
@@ -297,35 +299,35 @@ def real_sh_tournier_from_index(m, n, theta, phi, legacy=True):
 
 
-def real_sh_descoteaux_from_index(m, n, theta, phi, legacy=True):
+def real_sh_descoteaux_from_index(m_values, l_values, theta, phi, legacy=True):
```
same here, see above

skoudoro reviewed (commented):
Thank you for this great work @itellaetxe. 

This PR is almost ready to go. See my additional comments.

is there any additional comments @CHrlS98 ? @karanphil ?  @arokem ? @ebrahimebrahim ?

skoudoro commented:
If not, I might edit this PR, let the CI's run and merge it today to be part of the release.

skoudoro commented:
Ok, all seems ok, merging!

Thank you @itellaetxe for this work

Thank you @arokem @CHrlS98 @karanphil @ebrahimebrahim  for the review/debate

itellaetxe commented:
Thank you @skoudoro for the help with those last decorators!
Also thank you for the patience and the good advice to everyone that participated in the discussion. I will try to do it better next time :)
